owltorrent-network/src/main/java/edu/rice/owltorrent/network/TorrentManager.java

```diff
@@ -270,6 +278,25 @@ public void run() {
     }
   }
 
+  private boolean requestFromLeecher(int index, PieceStatus pieceStatus, int blockIndex) {
+    boolean leecherFlag = false;
+    if (!indexToLeechers.containsKey(index)) return leecherFlag;
+
+    Set<Peer> leecherSet = indexToLeechers.get(index);
+    while (leecherSet.iterator().hasNext()) {
+      Peer leecher = leecherSet.iterator().next();
```

Comment:
Are you sure this works as intended? If you call leecherSet.iterator() each time, don't you generate a new iterator each time, which causes this to infinite loop? 

You can use for (Peer peer : leecherSet) instead...